### PR TITLE
Sessions

### PR DESCRIPTION
**Pull request by justincjohnson**
I walked through the code base overall with @softwareplumber and am using this PR as a syncing point between the two of us.  These changes primarily get push up to speed with my current spec knowledge.  After this I'll do the same for pull, followed by thinking through appropriate abstractions to support streaming, adding session cache, etc..

### CODE REVIEW

File: carmirror/car.go [hunk 1]
```diff
@@ -10,10 +10,12 @@ import (
 	"github.com/ipld/go-car"
 )
 
-func AddAllFromCarReader(ctx context.Context, bapi coreiface.BlockAPI, r io.Reader, progCh chan cid.Cid) (int, error) {
+func AddAllFromCarReader(ctx context.Context, bapi coreiface.BlockAPI, r io.Reader, progCh chan cid.Cid) (int, []cid.Cid, error) {
+	log.Debugf("AddAllFromCarReader")
+	cids := []cid.Cid{}
```
Comment (softwareplumber) [inline carmirror/car.go:15]: Could we allocate a capacity of at least len(rdr.Header.Roots) entries in the cid slice in order to speed things up?

File: carmirror/carmirror.go [hunk 2]
```diff
@@ -303,44 +358,72 @@ func NewPullHandler(cm *CarMirror) http.HandlerFunc {
...
+// GetLocalCids returns a unique list of `gocid.Cid`s underneath a given root CID, using an offline CoreAPI.
 // The root CID is included in the returned list.
 // In the case of an error, both the discovered CIDs thus far and the error are returned.
-func (cm *CarMirror) GetLocalCids(ctx context.Context, rootCidStr string) ([]cid.Cid, error) {
-	var cids []cid.Cid
-	rootCid, err := dag.ParseCid(rootCidStr)
-	if err != nil {
-		return nil, errors.Wrapf(err, "unable to parse root cid %s", rootCidStr)
+func (cm *CarMirror) GetLocalCids(ctx context.Context, rootCids []gocid.Cid) ([]gocid.Cid, error) {
+	var cids []gocid.Cid
```
Comment (softwareplumber) [inline carmirror/carmirror.go:365]: Again, could we set the capacity to be at least len(rootCids)?

File: dag/dag.go [hunk 3]
```diff
@@ -10,14 +10,17 @@ import (
 // Returns `nil, error` if any CID strings fail to parse.
 func ParseCids(cids []string) ([]cid.Cid, error) {
 	var rootCids []cid.Cid
```
Comment (softwareplumber) [inline dag/dag.go:12]: Preallocate :-)

**Review by softwareplumber — APPROVED** [review]
LGTM. There are a couple of non-critical comments,

File: carmirror/car.go (hunk 1 above)
Comment (justincjohnson) [reply carmirror/car.go]: Fabulous. Thank you!

**Pull request merged**